canary.go

```diff
@@ -49,58 +50,100 @@ func (c *Canary) SignalHandler() {
 			}
 			os.Exit(0)
 		case syscall.SIGHUP:
+			manifest, err := manifest.GetManifest(c.Config.ManifestURL, c.Config.DefaultSampleInterval)
+			if err != nil {
+				log.Fatal(err)
+			}
 			// Split reload logic into reloader() as to allow other things to trigger a manifest reload
-			c.ReloadChan <- true
+			c.ReloadChan <- manifest
 		}
 	}
 }
 
 func (c *Canary) reloader() {
 	if c.ReloadChan == nil {
-		c.ReloadChan = make(chan bool)
+		c.ReloadChan = make(chan manifest.Manifest)
 	}
 
-	for r := range c.ReloadChan {
-		if r {
-			// stop all running sensors
-			for _, sensor := range c.Sensors {
-				sensor.Stop()
+	for m := range c.ReloadChan {
+		stoppingSensors := []sensor.Sensor{}
+		for _, sensor := range c.Sensors {
+			found := false
+			for _, newTarget := range m.Targets {
+				if newTarget.Hash == sensor.Target.Hash {
+					found = true
+				}
 			}
-			for _, sensor := range c.Sensors {
-				<-sensor.IsStopped
+			if !found {
+				sensor.Stop()
+				stoppingSensors = append(stoppingSensors, sensor)
 			}
 
-			// get an updated manifest.
-			manifest, err := manifest.GetManifest(c.Config.ManifestURL, c.Config.DefaultSampleInterval)
-			if err != nil {
-				log.Fatal(err)
-			}
-			c.Manifest = manifest
-			if c.Config.RampupSensors {
-				c.Manifest.GenerateRampupDelays(c.Config.DefaultSampleInterval)
-			}
-			// Start new sensors:
-			c.startSensors()
 		}
+		for _, sensor := range stoppingSensors {
+			<-sensor.StopNotifyChan
+		}
+
+		c.Manifest = m
+		if c.Config.RampupSensors {
+			c.Manifest.GenerateRampupDelays(c.Config.DefaultSampleInterval)
+		}
+		// Start new sensors:
+		c.startSensors()
 	}
 }
 
 func (c *Canary) startSensors() {
+	oldSensors := c.Sensors
 	c.Sensors = []sensor.Sensor{} // reset the slice
 
 	// spinup a sensor for each target
 	for index, target := range c.Manifest.Targets {
-		sensor := sensor.Sensor{
-			Target:    target,
-			C:         c.OutputChan,
-			Sampler:   sampler.New(),
-			StopChan:  make(chan int, 1),
-			IsStopped: make(chan bool),
-			IsOK:      false,
+		found := false
```

Comment:
this looks a bit weird. i think this does:

* blank the sensor list, then check each sensor in the manifest
** if the new sensor exists in the old list and hasn't changed, just add it
** else create a new sensor with the manifest data

something like this: https://gist.github.com/beejeebus/c01fe4dd2aae4ecb1372